storage/storage-interface/src/async_proof_fetcher.rs

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{proof_fetcher::ProofFetcher, DbReader};
+
+use anyhow::Result;
+use aptos_crypto::{hash::CryptoHash, HashValue};
+use aptos_types::{
+    proof::SparseMerkleProof,
+    state_store::{state_key::StateKey, state_value::StateValue},
+    transaction::Version,
+};
+use crossbeam_channel::{unbounded, Receiver, Sender};
+use num_cpus;
+use once_cell::sync::Lazy;
+use std::{
+    cmp::min,
+    collections::HashMap,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc,
+    },
+};
+
+static IO_POOL: Lazy<rayon::ThreadPool> = Lazy::new(|| {
+    rayon::ThreadPoolBuilder::new()
+        // TODO(grao): Do more analysis to tune this magic number.
+        .num_threads(min(256, 4 * num_cpus::get()))
+        .thread_name(|index| format!("proof_reader_{}", index))
+        .build()
+        .unwrap()
+});
+
+struct Proof {
+    state_key_hash: HashValue,
+    proof: SparseMerkleProof,
+}
+
+pub struct AsyncProofFetcher {
+    reader: Arc<dyn DbReader>,
+    data_sender: Sender<Proof>,
+    data_receiver: Receiver<Proof>,
+    num_proofs_to_read: AtomicUsize,
+}
+
+impl AsyncProofFetcher {
+    pub fn new(reader: Arc<dyn DbReader>) -> Self {
+        let (data_sender, data_receiver) = unbounded();
+
+        Self {
+            reader,
+            data_sender,
+            data_receiver,
+            num_proofs_to_read: AtomicUsize::new(0),
+        }
+    }
+
+    // Returns all read proofs. This is only expected to be called after all reads being scheduled,
+    // in a single thread.
+    fn finish_and_read_proofs(&self) -> HashMap<HashValue, SparseMerkleProof> {
```

Comment:
"read_proofs" obviously happens before "finish", right? :trollface: 

I'd just call it `finish()`, etc or maybe "read_async()" -> "schedule() / queue()" and "finish()" -> "wait()"

Reply:
Done